<a href="https://colab.research.google.com/github/angelvaz20/Colab_files/blob/main/Lab_9_Multinomial.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
from sklearn.naive_bayes import MultinomialNB
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix
import pandas as pd
import numpy as np

In [ ]:
df = pd.read_csv("tweets.csv")
# Keep only the columns we need
df = df[["text", "target"]]

# Remove missing values just in case
df = df.dropna()

# Get the tweets and labels
tweets = df["text"]
labels = df["target"]
df

,text,target
0,"Communal violence in Bhainsa, Telangana. ""Ston...",1
1,Telangana: Section 144 has been imposed in Bha...,1
2,Arsonist sets cars ablaze at dealership https:...,1
3,Arsonist sets cars ablaze at dealership https:...,1
4,"""Lord Jesus, your love brings freedom and pard...",0
...,...,...
11365,Media should have warned us well in advance. T...,0
11366,i feel directly attacked 💀 i consider moonbin ...,0
11367,i feel directly attacked 💀 i consider moonbin ...,0
11368,"ok who remember ""outcast"" nd the ""dora"" au?? T...",0


In [ ]:
print("=== STEP 2: Dataset loaded ===")
print(f"Number of tweets: {len(tweets)}")
print()

=== STEP 2: Dataset loaded ===
Number of tweets: 11370



In [ ]:
vectorizer = CountVectorizer()

In [ ]:
X = vectorizer.fit_transform(tweets)

In [ ]:
print("=== STEP 3: Words converted to numbers ===")
print(f"Number of unique words found: {len(vectorizer.get_feature_names_out())}")
print()

=== STEP 3: Words converted to numbers ===
Number of unique words found: 27112



In [ ]:
model = MultinomialNB()
model.fit(X, labels)

MultinomialNB()

In [ ]:
from sklearn.model_selection import train_test_split
# Split the data (this is IMPORTANT for reliability)

X_train, X_test, y_train, y_test = train_test_split(
    tweets, labels, test_size=0.2, random_state=42
)

In [ ]:
# Vectorize AGAIN (fit only on training data)
X_train_vec = vectorizer.fit_transform(X_train)
X_test_vec = vectorizer.transform(X_test)

In [ ]:
# Train model on training data
model.fit(X_train_vec, y_train)

# Make predictions on TEST data
y_pred = model.predict(X_test_vec)

In [ ]:
# Results
# 1. Accuracy
accuracy = accuracy_score(y_test, y_pred)
print(f"Accuracy: {accuracy:.2f}")
print()

# 2. Classification Report
print("Classification Report:")
print(classification_report(y_test, y_pred))
print()

Accuracy: 0.89

Classification Report:
              precision    recall  f1-score   support

           0       0.91      0.96      0.94      1878
           1       0.76      0.55      0.64       396

    accuracy                           0.89      2274
   macro avg       0.83      0.76      0.79      2274
weighted avg       0.88      0.89      0.88      2274




In [ ]:
# Testing Model on new tweets not seen before
new_tweets = [
    "Massive fire destroys several homes in California",
    "I love this new song so much",
    "Earthquales shakes downtown area and damages buldings",
    "Had a great lunch with my friends today",
    "Flood warning issued after heavy rain in the city ",
    "I loved reading those books from the book club"

]
# Convert new tweets into numbers using the SAME vectorizer
X_new = vectorizer.transform(new_tweets)

# Ask the model to make predictions
predictions = model.predict(X_new)

# Get probability scores
probabilities = model.predict_proba(X_new)

In [ ]:
print("=== 3Predictions on new tweets ===")
print()

for i, tweet in enumerate(new_tweets):
    # predictions[i] is either 1 (disaster) or 0 (not disaster)
    result = "DISASTER" if predictions[i] == 1 else "NOT DISASTER"

    # probabilities[i][0] = probability of NOT disaster
    # probabilities[i][1] = probability of disaster
    prob_disaster = probabilities[i][1] * 100
    prob_notdisaster = probabilities[i][0] * 100

    print(f"Tweet: '{tweet}'")
    print(f"  --> Prediction      : {result}")
    print(f"  --> Disaster chance : {prob_disaster:.1f}%")
    print(f"  --> Not disaster    : {prob_notdisaster:.1f}%")
    print()

=== Predictions on new tweets ===

Tweet: 'Massive fire destroys several homes in California'
  --> Prediction      : DISASTER
  --> Disaster chance : 99.7%
  --> Not disaster    : 0.3%

Tweet: 'I love this new song so much'
  --> Prediction      : NOT DISASTER
  --> Disaster chance : 0.0%
  --> Not disaster    : 100.0%

Tweet: 'Earthquales shakes downtown area and damages buldings'
  --> Prediction      : DISASTER
  --> Disaster chance : 69.5%
  --> Not disaster    : 30.5%

Tweet: 'Had a great lunch with my friends today'
  --> Prediction      : NOT DISASTER
  --> Disaster chance : 0.2%
  --> Not disaster    : 99.8%

Tweet: 'Flood warning issued after heavy rain in the city '
  --> Prediction      : DISASTER
  --> Disaster chance : 99.5%
  --> Not disaster    : 0.5%

Tweet: 'I loved reading those books from the book club'
  --> Prediction      : NOT DISASTER
  --> Disaster chance : 0.0%
  --> Not disaster    : 100.0%



# Summary:
This dataset shows over 11,000 tweets associated with disaster keywords like “crash”, “quarantine”, and “bush fires” as well as the location and keyword itself. What we did was make sure that the dataset focused on the text (tweet) and the target (either 1 = disaster or 0 = no disaster). We then used CountVectorizers which converts words into numbers because numbers are better understood than words in this context. We then counted how many unique words appeared. After that we created and trained the Naive Bayes model basically showing the model our dataset so it can learn from it. From there we measured the reliability and the classification report which resulted in an 89% reliability score. We then created new tweets that were new to the model in order to see if the model can see if the tweet was a disaster or wasn't a disaster. From those tweets we made results where the model predicts if the tweet is about a disaster or not a disaster from the new tweets given.